Add documentation for UTF-8 strict mode (#8596)
grafana/mimir: docs/sources/mimir/references/architecture/components/alertmanager.md

Point the address to Alertmanager’s API.
You can configure Alertmanager’s API prefix via the `-http.alertmanager-http-prefix` flag, which defaults to `/alertmanager`.
For example, if Alertmanager is listening at `http://mimir-alertmanager.namespace.svc.cluster.local` and it is using the default API prefix, set `-ruler.alertmanager-url` to `http://mimir-alertmanager.namespace.svc.cluster.local/alertmanager`.

## Enable UTF-8

> In a wide effort to support alerts from [OpenTelemetry](https://opentelemetry.io/) (OTel) data, [Prometheus Alertmanager](https://prometheus.io/docs/alerting/alertmanager/) has added support for UTF-8. This is supported as an opt-in feature for the Grafana Mimir Alertmanager in Mimir versions 2.12 and later, and will be changed to opt-out in future versions of Mimir.

Author: Oh OK. I think we want some language here that encourages users to opt-in before it is changed into an opt-out, because as is written further down, some users might have issues with some of the breaking changes, and we don't want it to be a surprise for them. How best should we word it?

> In a wide effort to support alerts from [OpenTelemetry](https://opentelemetry.io/) (OTel) data, [Prometheus Alertmanager](https://prometheus.io/docs/alerting/alertmanager/) has added support for UTF-8. This is supported as an opt-in feature for the Grafana Mimir Alertmanager in Mimir versions 2.12 and later, and will be changed to opt-out in future versions of Mimir.

Author: I'll remove it, and we can add it back later if we need to 👍

> In a wide effort to support alerts from [OpenTelemetry](https://opentelemetry.io/) (OTel) data, [Prometheus Alertmanager](https://prometheus.io/docs/alerting/alertmanager/) has added support for UTF-8. This is supported as an opt-in feature for the Grafana Mimir Alertmanager in Mimir versions 2.12 and later, and will be changed to opt-out in future versions of Mimir.

Reviewer: I think it's still best to just leave it as opt-in for now while still encouraging people to do it.

> In a wide effort to support alerts from [OpenTelemetry](https://opentelemetry.io/) (OTel) data, [Prometheus Alertmanager](https://prometheus.io/docs/alerting/alertmanager/) has added support for UTF-8. This is supported as an opt-in feature for the Grafana Mimir Alertmanager in Mimir versions 2.12 and later, and will be changed to opt-out in future versions of Mimir.

Reviewer: We can encourage people to opt-in without mentioning that the setting will change in future releases.

If this is a new Mimir installation, we recommend enabling support for UTF-8 before creating any tenant configurations. You can do this by changing [`utf8-strict-mode-enabled`]({{< relref "./../../../configure/configuration-parameters#alertmanager" >}}) to true.

If this is an existing Mimir installation, there are a number of breaking changes that might affect existing tenant configurations. We recommend following the instructions below to ensure all existing tenant configurations are compatible with UTF-8 before enabling it.

### What are the breaking changes?

> In order to support UTF-8, Alertmanager has added a new parser for label matchers (often abbreviated as matchers), which has a number of breaking changes. While most matchers will be forward-compatible, some will not.

Author: Is there a different phrase we can use instead? I took this from the Prometheus docs, but we want to communicate that most users can expect to be unimpacted (based on the the numbers we have seen in Grafana Cloud), although it's still important to follow the process to make sure.

{{< admonition type="note" >}}
If you are unfamiliar with what matchers are or how they are used in a tenant configuration, you can find more information about them in the [Prometheus Alertmanager documentation](https://prometheus.io/docs/alerting/latest/configuration).
{{< /admonition >}}

> Grafana Mimir provides a number of tools to help you identify if any existing tenant configurations would be affected by these breaking changes, and migrate any affected tenant configurations in a way that is backwards compatible, does not change the behavior of existing matchers, and works even in Mimir installations that do not have UTF-8 enabled.

Reviewer: ```suggestion
Grafana Mimir provides a number of tools to help you identify whether any existing tenant configurations are affected by these breaking changes and to migrate any affected tenant configurations in a way that is backwards-compatible, doesn't change the behavior of existing matchers, and works even in Mimir installations that do not have UTF-8 enabled.
```

> ### Identifying affected tenant configurations

Reviewer: ```suggestion
### Identify affected tenant configurations
```

To identify affected tenant configurations, take the following steps:

> 1. Make sure Mimir is running version 2.12 or newer with debug-level logging enabled.

Reviewer: ```suggestion
1. Make sure Mimir is running version 2.12 or later with debug-level logging enabled.
```

> 2. To identify any tenants configurations that are incompatible with UTF-8 (meaning the tenant configuration would fail to load and the [fallback configuration](#fallback-configuration) would be used instead), search Mimir server logs for lines containing `Alertmanager is moving to a new parser for labels and matchers, and this input is incompatible`. Each log line includes the invalid matcher from the tenant configuration and the ID of the affected tenant. For example:

Reviewer: ```suggestion
2. To identify any tenant configurations that are incompatible with UTF-8 (meaning the tenant configuration fails to load and the [fallback configuration](#fallback-configuration) is used instead), search Mimir server logs for lines containing `Alertmanager is moving to a new parser for labels and matchers, and this input is incompatible`. Each log line includes the invalid matcher from the tenant configuration and the ID of the affected tenant. For example:
```

    ```

>     msg="Alertmanager is moving to a new parser for labels and matchers, and this input is incompatible. Alertmanager has instead parsed the input using the classic matchers parser as a fallback. To make this input compatible with the UTF-8 matchers parser please make sure all regular expressions and values are double-quoted. If you are still seeing this message please open an issue." input="foo=" origin=api err="end of input: expected label value" suggestion="foo=\"\"" user="1"

Reviewer: ```suggestion
    msg="Alertmanager is moving to a new parser for labels and matchers, and this input is incompatible. Alertmanager has instead parsed the input using the classic matchers parser as a fallback. To make this input compatible with the UTF-8 matchers parser, make sure all regular expressions and values are double-quoted. If you still see this message, open an issue." input="foo=" origin=api err="end of input: expected label value" suggestion="foo=\"\"" user="1"
```

Reviewer: We try to avoid saying "please" in the docs.

Author: Can't change this, it's a literal error message from the software.

    ```

>     In this example you can see that the tenant with User ID `1` has an incompatible matcher in their tenant configuration `foo=`, and should to be changed to the suggestion `foo=""`.

Reviewer: ```suggestion
    In this example, the tenant with the  `1` User ID has an incompatible matcher in its `foo=` tenant configuration and should be changed to the suggestion, `foo=""`.
```

> 3. To identify any tenant configurations that are compatible with UTF-8, but the behavior of some of their matchers could change when enabling it (this is called disagreement), search Mimir server logs for lines containing `Matchers input has disagreement`. Disagreement occurs when a matcher is valid, but due to adding support for UTF-8, it can behave differently when UTF-8 is enabled.

Reviewer: ```suggestion
3. To identify any tenant configurations that are compatible with UTF-8 but have some matchers that might change in behavior when its enabled, search Mimir server logs for lines containing `Matchers input has disagreement`. Disagreement occurs when a matcher is valid but behaves differently when UTF-8 is enabled.
```

    ```
    msg="Matchers input has disagreement" input="foo=\"\\xf0\\x9f\\x99\\x82\"" user="1"
    ```

{{< admonition type="note" >}}

> It is possible for a tenant configuration to be both incompatible with UTF-8 and have disagreement, as an individual tenant configuration can have 10s or even 100s of matchers. For example, in different routes, inhibition rules, etc.

Reviewer: ```suggestion
It's possible for a tenant configuration to be both incompatible with UTF-8 and have disagreement, as an individual tenant configuration can have 10s or even 100s of matchers. 
```

Reviewer: I think we can remove the last sentence from this admonition.

{{< /admonition >}}

> ### Fixing identified tenant configurations

Reviewer: ```suggestion
### Fix tenant configurations
```

To fix any identified tenant configurations, take the following steps:

> 1. Tenant configurations that are incompatible with UTF-8 can be fixed by using the [`migrate-utf8`]({{< relref "./../../../manage/tools/mimirtool#migrate-alertmanager-configuration-for-utf-8-in-mimir-212-and-later" >}}) command in mimirtool. This [command]({{< relref "./../../../manage/tools/mimirtool#migrate-alertmanager-configuration-for-utf-8-in-mimir-212-and-later" >}}) can migrate existing tenant configurations in a way that is backwards compatible, does not change the behavior of existing matchers, and works even in Mimir installations that do not have UTF-8 enabled. If using mimirtool is not possible, a tenant configuration can be edited by hand, applying each  suggestion from the Mimir server logs.

Reviewer: ```suggestion
1. Use the [`migrate-utf8`]({{< relref "./../../../manage/tools/mimirtool#migrate-alertmanager-configuration-for-utf-8-in-mimir-212-and-later" >}}) command in mimirtool to fix any tenant configurations that are incompatible with UTF-8. This [command]({{< relref "./../../../manage/tools/mimirtool#migrate-alertmanager-configuration-for-utf-8-in-mimir-212-and-later" >}}) can migrate existing tenant configurations in a way that is backwards-compatible, doesn't change the behavior of existing matchers, and works even in Mimir installations that don't have UTF-8 enabled. If you cannot use mimirtool, you can edit tenant configurations by hand through applying each suggestion from the Mimir server logs.
```

> 2. Tenant configurations that have disagreement must be looked at on a case by case basis as depending on the nature of the disagreement, a matcher with disagreement might not need fixing. For example `\xf0\x9f\x99\x82` is the byte sequence for the 🙂 emoji. If the intention is to match a literal 🙂 emoji then no change is required. However, if the intention is to match the literal `\xf0\x9f\x99\x82` then the matcher should be changed to `qux="\\xf0\\x9f\\x99\\x82"`.

Reviewer: ```suggestion
2. You must look at tenant configurations that have disagreement on a case-by-case basis. Depending on the nature of the disagreement, you might not need to fix a matcher with disagreement. For example `\xf0\x9f\x99\x82` is the byte sequence for the 🙂 emoji. If the intention is to match a literal 🙂 emoji, then no change is required. However, if the intention is to match the literal `\xf0\x9f\x99\x82`, then you need to change the matcher to `qux="\\xf0\\x9f\\x99\\x82"`.
```

{{< admonition type="note" >}}

> It is rare to find cases of disagreement in a tenant configuration as most tenants do not need to match alerts that contain literal UTF-8 bytes sequences in their labels.

Reviewer: ```suggestion
It's rare to find cases of disagreement in a tenant configuration, as most tenants don't need to match alerts that contain literal UTF-8 byte sequences in their labels.
```

{{< /admonition >}}

### Final steps

> 1. Once all affected tenant configurations have been identified and fixed, check the Mimir server logs a second time to make sure no tenant configurations have been missed.

Reviewer: ```suggestion
1. After identifying and fixing all affected tenant configurations, check the Mimir server logs again to make sure you haven't missed any tenant configurations.
```

> 2. You can now enable UTF-8 by changing [`utf8-strict-mode-enabled`]({{< relref "./../../../configure/configuration-parameters#alertmanager" >}}) to true.

Reviewer: ```suggestion
2. You can now enable UTF-8 by setting [`utf8-strict-mode-enabled`]({{< relref "./../../../configure/configuration-parameters#alertmanager" >}}) to `true`.
```

> 3. Any incompatible tenant configurations will fail to load. You can make sure all tenant configurations are working by searching the Mimir server logs for lines containing `error applying config`, or the `alertmanager_config_last_reload_successful` gauage for `0`.

Reviewer: ```suggestion
3. Any incompatible tenant configurations fail to load. You can make sure all tenant configurations work by searching the Mimir server logs for lines containing `error applying config`, or the `alertmanager_config_last_reload_successful` gauge for `0`.
```
